server/compactor/src/main/java/org/apache/accumulo/compactor/Compactor.java

```diff
@@ -0,0 +1,889 @@
+
+  protected static final CompactionJobHolder JOB_HOLDER = new CompactionJobHolder();
+
+  private final GarbageCollectionLogger gcLogger = new GarbageCollectionLogger();
+  private final UUID compactorId = UUID.randomUUID();
+  private final AccumuloConfiguration aconf;
+  private final String queueName;
+  private final AtomicReference<CompactionCoordinator.Client> coordinatorClient =
+      new AtomicReference<>();
+  protected final AtomicReference<ExternalCompactionId> currentCompactionId =
+      new AtomicReference<>();
+
+  private SecurityOperation security;
+  private ServiceLock compactorLock;
+  private ServerAddress compactorAddress = null;
+
+  // Exposed for tests
+  protected volatile Boolean shutdown = false;
+
+  protected Compactor(CompactorServerOpts opts, String[] args) {
+    super("compactor", opts, args);
+    queueName = opts.getQueueName();
+    aconf = getConfiguration();
+    setupSecurity();
+    var schedExecutor = ThreadPools.createGeneralScheduledExecutorService(aconf);
+    startGCLogger(schedExecutor);
+    startCancelChecker(schedExecutor, TIME_BETWEEN_CANCEL_CHECKS);
+    printStartupMsg();
+  }
+
+  protected Compactor(CompactorServerOpts opts, String[] args, AccumuloConfiguration conf) {
+    super("compactor", opts, args);
+    queueName = opts.getQueueName();
+    aconf = conf;
+    setupSecurity();
+    var schedExecutor = ThreadPools.createGeneralScheduledExecutorService(aconf);
+    startGCLogger(schedExecutor);
+    startCancelChecker(schedExecutor, TIME_BETWEEN_CANCEL_CHECKS);
+    printStartupMsg();
+  }
+
+  protected void setupSecurity() {
+    getContext().setupCrypto();
+    security = AuditedSecurityOperation.getInstance(getContext());
+  }
+
+  protected void startGCLogger(ScheduledThreadPoolExecutor schedExecutor) {
+    schedExecutor.scheduleWithFixedDelay(() -> gcLogger.logGCInfo(getConfiguration()), 0,
+        TIME_BETWEEN_GC_CHECKS, TimeUnit.MILLISECONDS);
+  }
+
+  protected void startCancelChecker(ScheduledThreadPoolExecutor schedExecutor,
+      long timeBetweenChecks) {
+    schedExecutor.scheduleWithFixedDelay(() -> checkIfCanceled(), 0, timeBetweenChecks,
+        TimeUnit.MILLISECONDS);
+  }
+
+  protected void checkIfCanceled() {
+    TExternalCompactionJob job = JOB_HOLDER.getJob();
+    if (job != null) {
+      try {
+        var extent = KeyExtent.fromThrift(job.getExtent());
+        var ecid = ExternalCompactionId.of(job.getExternalCompactionId());
+
+        TabletMetadata tabletMeta =
+            getContext().getAmple().readTablet(extent, ColumnType.ECOMP, ColumnType.PREV_ROW);
+        if (tabletMeta == null || !tabletMeta.getExtent().equals(extent)
+            || !tabletMeta.getExternalCompactions().containsKey(ecid)) {
+          // table was deleted OR tablet was split or merged OR tablet no longer thinks compaction
+          // is running for some reason
+          LOG.info("Cancelling compaction {} that no longer has a metadata entry at {}", ecid,
```

Comment:
@ctubbsii I Am currently working through a complete review of the code.  You had mentioned having a very generic per tserver plugin to generally handle external compactions.  Thinking about that when looking at this code, it cancels a running external compaction for a tablet that was deleted (split/merge/table deletion).  To handle this case in a highly pluggable environment, this functionality would need to be present in some form in the SPI.   This would be unrelated to any pluggable component on a tserver, its more related to running external compactions.  The tablet for a running external compaction could be unload and then the tablet deleted, no tserver would be aware of this.